tsdb/engine/tsm1/engine.go

```diff
@@ -612,73 +612,146 @@ func (e *Engine) writeFileToBackup(name string, shardRelativePath, fullPath stri
 // Only files that match basePath will be copied into the directory. This obtains
 // a write lock so no operations can be performed while restoring.
 func (e *Engine) Restore(r io.Reader, basePath string) error {
+	return e.overlay(r, basePath, false)
+}
+
+// Import reads a tar archive generated by Backup() and adds each
+// file matching basePath as a new TSM file.  This obtains
+// a write lock so no operations can be performed while Importing.
+func (e *Engine) Import(r io.Reader, basePath string) error {
+	return e.overlay(r, basePath, true)
+}
+
+// restore reads a tar archive generated by Backup() and adds each file
+// from the archive matching basePath to the shard.
+// If asNew is true, each file will be installed as a new TSM file even if an
+// existing file with the same name in the backup exists.
+func (e *Engine) overlay(r io.Reader, basePath string, asNew bool) error {
 	// Copy files from archive while under lock to prevent reopening.
-	if err := func() error {
+	newFiles, err := func() ([]string, error) {
 		e.mu.Lock()
 		defer e.mu.Unlock()
 
+		var newFiles []string
 		tr := tar.NewReader(r)
 		for {
-			if err := e.readFileFromBackup(tr, basePath); err == io.EOF {
+			if fileName, err := e.readFileFromBackup(tr, basePath, asNew); err == io.EOF {
 				break
 			} else if err != nil {
-				return err
+				return nil, err
+			} else if fileName != "" {
+				newFiles = append(newFiles, fileName)
 			}
 		}
 
-		return syncDir(e.path)
-	}(); err != nil {
+		if err := syncDir(e.path); err != nil {
+			return nil, err
+		}
+
+		if err := e.FileStore.Replace(nil, newFiles); err != nil {
+			return nil, err
+		}
+		return newFiles, nil
+	}()
+
+	if err != nil {
 		return err
 	}
 
+	// Load any new series keys to the index
+	readers := make([]chan seriesKey, 0, len(newFiles))
+	for _, f := range newFiles {
+		ch := make(chan seriesKey, 1)
+		readers = append(readers, ch)
+
+		if strings.HasSuffix(f, ".tmp") {
+			f = strings.TrimSuffix(f, ".tmp")
+		}
+
+		fd, err := os.Open(f)
+		if err != nil {
+			return err
+		}
+
+		r, err := NewTSMReader(fd)
+		if err != nil {
+			return err
+		}
+		defer r.Close()
+
+		go func(c chan seriesKey, r *TSMReader) {
+			n := r.KeyCount()
+			for i := 0; i < n; i++ {
+				key, typ := r.KeyAt(i)
+				c <- seriesKey{key, typ}
+			}
+			close(ch)
+		}(ch, r)
```

Comment:
The `close(ch)` references the `ch` outside the closure while the arg passed in is called `c`. I don't think it should cause an issue since `ch` is defined inside the `for` loop but I think it could lead to subtle bugs if the `ch` definition is moved outside the loop at some point.